### PR TITLE
Clean up Sink implementation on Framed

### PR DESCRIPTION
**Pull request by danburkert**
This commit makes a few changes to the Sink implementation on Framed:

* Backpressure is implemented for `start_send`. If the write buffer is
  over 8KiB and can't be flushed, no new items are accepted.
* 0 length writes to the upstream transport are translated into a
  `WriteZero` error, as with `io::Write::write_all`. `write_all` checks
  for and ignores `Interrupted` errors, but I do not think this is
  necessary for non-blocking writes.
* Potentially fewer calls to prefix truncate the write buffer in
  `poll_complete`.
* In `poll_complete`, the upstream transport is not flushed until
  *after* writing the entire write buffer.

### CODE REVIEW

File: src/io/frame.rs [hunk 1]
```diff
@@ -357,27 +357,49 @@ impl<T: Io, C: Codec> Sink for Framed<T, C> {
...
+                    self.wr.drain(..offset);
+                    if error.kind() == io::ErrorKind::WouldBlock {
+                        return Ok(Async::NotReady);
+                    } else {
+                        return Err(error);
+                    }
+                }
             }
+        }
+        self.wr.drain(..offset);
```
Comment (alexcrichton) [inline src/io/frame.rs]: The major difference between this loop and the previous loop seems to be deferring the call to `drain` until the very end, which seems to be an optimization? Initial testing I've done ended up never showing that improvement in profiles, so I'm curious if you were able to get speedups with a change like this?

**Comment by alexcrichton** [comment]
Looks good to me, thanks!

File: src/io/frame.rs (hunk 1 above)
Comment (danburkert) [reply src/io/frame.rs]: I haven't tested, but I  would only expect it to show up if the upstream channel has a small buffer (unlike Tcp).  The main thing here was fixing the lack of backpressure and potential infinite loop on write zero - I can simplify this back to the way it was if you want.

Comment (alexcrichton) [reply src/io/frame.rs]: Ok let's stick to that for now. I'm totally fine optimizing this loop but I'd just prefer to have some numbers to see where it matters first.

**Comment by danburkert** [comment]
Removed the drain coalescing.

**Comment by alexcrichton** [comment]
Thanks!

**Pull request merged**